### ISSUE TITLE
plugin hbs stopped working for 0.20

### ISSUE THREAD

**Issue by AlbertWillemsen-Centric** [systemjs/systemjs#1658]
Getting:
`   TypeError: Cannot read property 'meta' of undefined
       at addDefaultExtension (C:\test\node_modules\systemjs\src\resolve.js:502:27)
       at applyPackageConfigSync (C:\test\node_modules\systemjs\src\resolve.js:551:25)
       at packageResolveSync (C:\test\node_modules\systemjs\src\resolve.js:244:10)
       at doMapSync (C:\test\node_modules\systemjs\src\resolve.js:574:29)
       at SystemJSLoader$1.packageResolveSync (C:\test\node_modules\systemjs\src\resolve.js:221:20)
       at SystemJSLoader$1.normalizeSync (C:\test\node_modules\systemjs\src\resolve.js:178:29)
       at SystemJSLoader$1.translate (file:///C:/test/src/vendor/github/davis/plugin-hbs@1.2.3/hbs.js!transpiled:6:44)
       at C:\test\node_modules\systemjs\src\instantiate.js:210:41`

**Comment by guybedford** [comment]
Odd it seems to be working ok for me. Perhaps post an issue to the Handlebars project. Also it would help to know your replication, because I just replicated it my side and didn't get any issues.

**Comment by AlbertWillemsen-Centric** [comment]
I'm using es2015 module syntax:
`import "test.hbs"`
and:
System.import("test.hbs")
Not the ! syntax.

And configured jspm config (for transpilded bundled version):
```
meta: {
	"*.hbs": {
		"defaultExtension": false
	}
```
and configured jspm config dev (for transpiling dev on the fly):
```
meta: {
	"*.hbs": {
			"loader": "hbs"
	}
```

**Comment by belackriv** [comment]
Getting the same thing: 
```
system.js:4 Uncaught (in promise) Error: Cannot read property 'meta' of undefined
  Instantiating https://localhost:8443/assets/lib/appLayoutView.hbs!https://localhost:8443/assets/jspm_packages/github/davis/plugin-hbs@1.2.3/hbs.js
  Loading https://localhost:8443/assets/lib/appLayoutView.js
  Loading https://localhost:8443/assets/lib/app.js
  Loading https://localhost:8443/assets/lib/main.js
  Loading lib/main.js
    at oe (system.js:4)
    at ie (system.js:4)
    at Q (system.js:4)
    at se (system.js:4)
    at Ve.Q (system.js:4)
    at Ve.X (system.js:4)
    at Ve.translate (hbs.js:7)
    at system.js:4
    at <anonymous>
﻿```

**Comment by nkbrs** [comment]
Same issue here, this forced me to return to system.js `0.19.41 `.

`0.20.14` produces the above error when using [hbs plugin](https://github.com/davis/plugin-hbs).

Reproduction steps: 

1. use SystemJS version `0.20.14`.
2. Install the hbs plugin via jspm: `jspm install hbs`
3. Configure `jspm.config.js` to support loading of `.hbs` files without `!` while doing imports:

```
SystemJS.config({
packages: {
    "app": {
      "meta": {
        "*.hbs": {
          "loader": "hbs"
        }
      }
    }
  }
});
```

4. Use the following code imported via SystemJS:

```
import render from './template.hbs';

console.log(render({title: "Hello World"}));
```

5. Contents of `template.hbs`

```
<p>{{title}}</p>
```

By the glimpse of it, this might be the issue on the plugin's side rather than SystemJS's, but it appears there have been changes between 0.19 and 0.20 that break this plugin. Admittedly, I didn't dig too much as I'm not a plugin developer, but a guy who's got stuff to work on. If I get the time, I'll gladly report what and why it breaks, possibly include a fix too. In the mean time, if anyone manages to sort this one out - that'd be awesome.

Cheers guys.

**Comment by AlbertWillemsen-Centric** [comment]
Hi Guy,

I've managed to find the bug.
I'll be submitting a pull shortly.

With kind regards

**Comment by frenkie** [comment]
Hi! I've just been looking into the same error and was able to pinpoint it to another (earlier) spot, resulting in an undefined `loader` parameter of `applyPackageConfigSync` which then get's passed on to `addDefaultExtension` and triggers this 'meta undefined' issue. 

Although the above pull request fixes `applyPackageConfigSync` by swapping `loader` for `config` (and `addDefaultExtension`'s signature also implies using `config` ) the origin lies in [doMapSync](https://github.com/systemjs/systemjs/blob/master/src/resolve.js#L563)

```
function doMapSync (loader, config, pkg, pkgKey, mapMatch, path, metadata, skipExtensions) {
  if (path[path.length - 1] === '/')
    path = path.substr(0, path.length - 1);
  var mapped = pkg.map[mapMatch];

  if (typeof mapped === 'object')
    throw new Error('Synchronous conditional normalization not supported sync normalizing ' + mapMatch + ' in ' + pkgKey);

  if (!validMapping(mapMatch, mapped, path) || typeof mapped !== 'string')
    return;

  return packageResolveSync.call(this, config, mapped + path.substr(mapMatch.length), pkgKey + '/', metadata, metadata, skipExtensions);
}
```

where `packageResolveSync.call` get's called with `this` as the function's context, BUT, `doMapSync` is only called directly throughout the code, never through `apply` or `call` so it should use it's own `loader` parameter to pass on to `packageResolveSync`. Function [doMap](https://github.com/systemjs/systemjs/blob/master/src/resolve.js#L614) actually does this right.

I will write an extra pull request for this issue if you guys don't mind.

**Comment by guybedford** [comment]
Thanks, released in 0.20.18.
